File: Ex0709_BinaryTreeNodeDistance/TreeNodeDist.cpp
#include <iostream>
#include <stack>
using namespace std;

typedef int ElementType;

typedef struct TreeNode* PtrToNode;
typedef PtrToNode Tree;

struct TreeNode
{
	ElementType data;
	Tree Left;
	Tree Right;
	Tree Parent;
};

Tree CreateTree(ElementType x)
{
	Tree T = new TreeNode;
	T->data = x;
	T->Left = nullptr;
	T->Right = nullptr;
	T->Parent = nullptr;
	return T;
}

void DelTree(PtrToNode T)
{
	if (T == nullptr)
	{
		return;
	}
	DelTree(T->Left);
	DelTree(T->Right);
	delete T;
}

// Insert a node into Tree
// Direction=0 means insert into left
// Direction=1 means insert into right
PtrToNode Insert(ElementType x, PtrToNode p, int direction)
{
	if (p == nullptr)
	{
		cout << "Error in inserting." << endl;
		return nullptr;
	}
	PtrToNode temp = CreateTree(x);
	temp->Parent = p;
	if (direction == 0)
	{
		p->Left = temp;
	}
	else
	{
		p->Right = temp;
	}
	return temp;
}

int Distance(PtrToNode a, PtrToNode b)
{
	if (a == nullptr || b == nullptr)
	{
		cout << "Error in computing distance." << endl;
		return -1;
	}
	
	stack<PtrToNode> a_ancestor;
	stack<PtrToNode> b_ancestor;
	PtrToNode current;
	for (current = a; current != nullptr; current = current->Parent)
	{
		a_ancestor.push(current);
	}
	for (current = b; current != nullptr; current = current->Parent)
	{
		b_ancestor.push(current);
	}

	while (a_ancestor.size() > 0 && b_ancestor.size() > 0 &&
		a_ancestor.top() == b_ancestor.top())
	{
		a_ancestor.pop();
		b_ancestor.pop();
	}

	return a_ancestor.size() + b_ancestor.size();
}

int main(void)
{
	Tree T = CreateTree(1);
	PtrToNode node1 = T;
	PtrToNode node2 = Insert(2, node1, 0);
	PtrToNode node3 = Insert(3, node1, 1);
	PtrToNode node4 = Insert(4, node2, 0);
	PtrToNode node5 = Insert(5, node2, 1);
	PtrToNode node6 = Insert(6, node3, 0);
	PtrToNode node7 = Insert(7, node3, 1);
	cout << "Distance between node4 and node7: " << Distance(node4, node7) << endl;
	cout << "Distance between node6 and node7: " << Distance(node6, node7) << endl;
	cout << "Distance between node3 and node7: " << Distance(node3, node7) << endl;

	DelTree(T);

	return 0;
}

File: Ex0709_BinaryTreeNodeDistance/TreeNodeDist.java
import java.util.Stack;

public class TreeNodeDist {
    public static void main(String[] args) {
        TreeNode<Integer> node1 = new TreeNode<>(1);
        TreeNode<Integer> node2 = node1.insert(2, 0);
        TreeNode<Integer> node3 = node1.insert(3, 1);
        TreeNode<Integer> node4 = node2.insert(4, 0);
        TreeNode<Integer> node5 = node2.insert(5, 1);
        TreeNode<Integer> node6 = node3.insert(6, 0);

        System.out.println("Distance between node4 and node6: "
                + node4.getDistance(node6));
        System.out.println("Distance between node4 and node5: "
                + node4.getDistance(node5));
        System.out.println("Distance between node4 and node2: "
                + node4.getDistance(node2));
    }
}

class TreeNode<E> {
    public E data;
    public TreeNode<E> left = null;
    public TreeNode<E> right = null;
    public TreeNode<E> parent = null;

    public TreeNode(E data) {
        this.data = data;
    }

    // Insert a node under a TreeNode
    // Direction=0 means insert into left
    // Direction=1 means insert into right
    public TreeNode<E> insert(E x, int direction) {
        TreeNode<E> newNode = new TreeNode<>(x);
        newNode.parent = this;

        if (direction == 0) {
            this.left = newNode;
        }
        else {
            this.right = newNode;
        }

        return newNode;
    }

    private Stack<TreeNode<E>> getAncestors() {
        Stack<TreeNode<E>> ancestors = new Stack<>();

        for (TreeNode<E> i = this; i != null; i = i.parent) {
            ancestors.push(i);
        }
        return ancestors;
    }

    public int getDistance(TreeNode<E> b) {
        if (b == null) {
            throw new NullPointerException();
        }

        Stack<TreeNode<E>> aAncestors = getAncestors();
        Stack<TreeNode<E>> bAncestors = b.getAncestors();

        while (aAncestors.size() > 0 && bAncestors.size() > 0
                && aAncestors.peek() == bAncestors.peek()) {
            aAncestors.pop();
            bAncestors.pop();
        }

        return aAncestors.size() + bAncestors.size();
    }
}
